PythonVritualenvOperator - Feature: Add support for private pypi repo packages in requirements

Bennyelg opened the pull request:
Currently PythonVritualenvOperator support requirements as a list of strings which indicates packages to install.
The current method is very naive and allow strict `pip install` and this doesn't allowed for example install packages from
private pypi repository. 

The code below, allow install packages from private repo as well from pypi server using fallback if not existing in the private repo.

Commit:
Adjust + basic test

Commit:
missing package for tests

mik-laj commented on airflow/operators/python.py:
```diff
@@ -308,6 +308,8 @@ def __init__(
         string_args: Optional[Iterable[str]] = None,
         templates_dict: Optional[Dict] = None,
         templates_exts: Optional[List[str]] = None,
```
Can you update docstring?

mik-laj commented on airflow/utils/python_virtualenv.py:
```diff
@@ -35,12 +35,20 @@ def _generate_virtualenv_cmd(tmp_dir: str, python_bin: str, system_site_packages
-    return cmd + requirements
+
+    private_cmd = [f'{tmp_dir}/bin/pip', 'install']
+    public_cmd = [f'{tmp_dir}/bin/pip', 'install' f'--trusted-host {repository_url} -i {index_url}']
```
Missing comma?

mik-laj commented on tests/operators/test_python.py:
```diff
@@ -856,6 +874,16 @@ def f():
+        try:
+            if not os.path.exists('/root/packages'):
+                os.makedirs('/root/packages')
+            subprocess.Popen(['pypi-server'])
```
Should we close this process to avoid side-effects?

Bennyelg replied:
I'll working on it now separately by figure out whats the best way to do it.

Bennyelg replied:
done

Commit:
PR Comments phase #1

Commit:
add pypi server stop after done

Commit:
Use Connection instance for pypi by add connection_id

Commit:
fallback is the pypi in case of not found in the private.

potiuk commented on tests/operators/test_python.py:
```diff
@@ -817,6 +822,31 @@ def f():
+            self._clean_stop_pypi_server()
+            self._initiate_pypi_server()
+            session.execute("delete from connection where conn_id = 'private_repo_test_conn';")
+            cn = Connection("private_repo_test_conn",
```
There is no need to create the connection. It's enough to patch the method with `AIRFLOW__CONN__NN` environment variables (https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#storing-a-connection-in-environment-variables  - also there is an example of how you can generate the connection url from the raw values - incluidng percent-encoding etc. https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#generating-a-connection-uri) . Another problem with this implementation is that it creates side-effect - the connection remains in the database after the test is run. If you termporarily patch env variable for the test, it will work faster and will not leave a side effect.

mik-laj replied:
See also docs about mocking connection: https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#mocking-variables-and-connections

potiuk commented on tests/operators/test_python.py:
```diff
@@ -817,6 +822,31 @@ def f():
+
+        with create_session() as session:
+            self._clean_stop_pypi_server()
+            self._initiate_pypi_server()
```
This could be pytest fixture or at the very least it should be either in try/finally or in setUp()/tearDown(). otherwise there is a risk the server stays running after test throws an exception.

potiuk commented on tests/operators/test_python.py:
```diff
@@ -856,6 +886,25 @@ def f():
+            if not os.path.exists('/root/packages'):
+                os.makedirs('/root/packages')
+            subprocess.Popen(['pypi-server'])
+            time.sleep(1.5)
```
How do we know the magic value of 1.5 is good ? Are you sure it is enough for all cases (running in CI etc?). That's a very bad idea to sleep in unit tests. It slows down the whole test suite. There should be a better way to check if the server is running. One option is to run an http request to it in a loop until it succeeds if there is no better way.

mik-laj commented on tests/operators/test_python.py:
```diff
@@ -817,6 +822,31 @@ def f():
+                system_site_packages=False,
+            )
+            session.commit()
+            self._clean_stop_pypi_server()
```
When an exception is thrown this server is not stopped and will still be able to create unexpected side effects. Whay do you think about fixtures? 
See: [Teardown/Cleanup (AKA Fixture finalization)](https://docs.pytest.org/en/6.2.x/fixture.html#teardown-cleanup-aka-fixture-finalization)

Bennyelg replied:
thanks will work on it

potiuk commented on setup.cfg:
```diff
@@ -160,6 +160,7 @@ install_requires =
     typing-extensions>=3.7.4;python_version<"3.8"
     unicodecsv>=0.14.1
     werkzeug~=1.0, >=1.0.1
+    pypiserver~=1.4
```
We only need pypi server for development. There is no need to make whole airflow depend on it in `install_requires`. Please move it to `[devel]` extra in `setup.py`.

potiuk commented on tests/operators/test_python.py:
```diff
@@ -856,6 +886,25 @@ def f():
+            pass
+
+    @staticmethod
+    def _clean_stop_pypi_server(port: bytes = b"8080"):
```
Too complex. running system command `killall  -9 pypi-server` does the job in one line of code.. And with the proper setup/teardown sequence, you should store the pid of the process started and try to kill it with SIGTERM rather than SIGKILL first anyway - this method should only be treated as a "safety net",

potiuk commented on airflow/utils/python_virtualenv.py:
```diff
@@ -35,12 +37,35 @@ def _generate_virtualenv_cmd(tmp_dir: str, python_bin: str, system_site_packages
+        port = con.port
+        host = con.host
+        if user:
+            index_url = f"{schema}://{user}:{password}@{host}:{port}/repository/python/simple"
```
There are few problems here:
* port is Optional [int] - it might be None - we need to provide sensible default if we want to hard-code : (or  handle None) case.
* "/repository/python/simple should not be hard-coded. I imagine registries can have different path. If anything the "/repository/python/simple" might be a default, but it should be overridable via extra in connection
* `https://` should be default for private registries. There are good reasons for private registry to be private and I cannot imagine a case where it would be http-based, leaving it essentially unprotected and with user/passwords sent in plain text.

potiuk commented on airflow/utils/python_virtualenv.py:
```diff
@@ -35,12 +37,35 @@ def _generate_virtualenv_cmd(tmp_dir: str, python_bin: str, system_site_packages
+            index_url = f"{schema}://{host}:{port}/repository/python/simple"
+        private_cmd = [f'{tmp_dir}/bin/pip',
+                       'install',
+                       f'--trusted-host', host,
```
I will revert to @uranusjr  here. He is our local PyPI expert (and pypi committer) so I trust 100% in his statements here.

Commit:
PR Comments alignemnt

uranusjr commented on airflow/utils/python_virtualenv.py:
```diff
@@ -35,12 +37,35 @@ def _generate_virtualenv_cmd(tmp_dir: str, python_bin: str, system_site_packages
+        password = con.get_password()
+        port = con.port or 8080
+        host = con.host
+        host_suffix = con.extra_dejson.get('host_suffix', 'repository/python/simple')
```
The default seems arbitrary (specific to Artifactory?) PEP 503 does not require the index to be hosted under a suffix, so IMO this should default to an empty string. Also, `host_suffix` feels like it should be appended after `host`, not after the port part. How about calling this `path`? (This is the terminology used by `urllib.parse`.

uranusjr commented on airflow/utils/python_virtualenv.py:
```diff
@@ -35,12 +37,35 @@ def _generate_virtualenv_cmd(tmp_dir: str, python_bin: str, system_site_packages
+        if user:
+            index_url = os.path.join(f"{schema}://{user}:{password}@{host}:{port}", host_suffix)
+        else:
+            index_url = os.path.join(f"{schema}://{host}:{port}", host_suffix)
```
Use `urllib.parse.urlunsplit` instead.

uranusjr commented on airflow/utils/python_virtualenv.py:
```diff
@@ -35,12 +37,35 @@ def _generate_virtualenv_cmd(tmp_dir: str, python_bin: str, system_site_packages
+        user = con.login
+        schema = con.schema or 'https'
+        password = con.get_password()
+        port = con.port or 8080
```
The default 8080 is also very arbitrary. The index is just an HTTP server, so this should default to have no explicit port.

uranusjr commented on airflow/utils/python_virtualenv.py:
```diff
@@ -35,12 +37,35 @@ def _generate_virtualenv_cmd(tmp_dir: str, python_bin: str, system_site_packages
+        private_cmd = [f'{tmp_dir}/bin/pip',
+                       'install',
+                       f'--index-url', index_url,
+                       f'--extra-index-url', 'https://pypi.org/simple'
```
> by set the priority of first check in the private repository, only if it does not exists go to the pypi.

[This is not how index works](https://stackoverflow.com/questions/44509415) (also see the pip issue linked in my previous comment). I recommend setting *only* `index_url` to the custom connection. If you really really must use `--extra-index-url` (there are a few use cases, but again, it is very unlikely yours is one of them), set the custom repository as extra and keep the default to PyPI.

mik-laj commented on tests/operators/test_python.py:
```diff
@@ -58,6 +63,22 @@
+        os.makedirs('/root/packages')
+    process = subprocess.Popen(['pypi-server', '-p', '8282'])
+    yield
+    _clean_stop_pypi_server(process.pid)
```
```suggestion
    process.kill()
```

Commit:
PR Comments alignement
1. use SplitUrl
2. extra-index-url only as an extra
3. change from url_suffix to path as a standard.
4. revoke port default

ashb commented on airflow/utils/python_virtualenv.py:
```diff
@@ -88,12 +124,15 @@ def prepare_virtualenv(
     :type requirements: List[str]
-    :return: Path to a binary file with Python in a virtual environment.
+    :param connection_id: The private repository connection_id
+     case there are private packages to install.
```
```suggestion
    :param connection_id: The connection ID of an http/https connection to use as the index url for installing
        pacakges from a private PyPi repository
```

uranusjr replied:
a private PyPI repository -> a private PEP 503 Simple Repository server, maybe? The term PyPI is confusing (officially it only refers to the “official” repository at PyPI.org)

potiuk replied:
Oh yeah. I was also wondering and struggled with the naming. Thanks for explaining :)

ashb commented on airflow/utils/python_virtualenv.py:
```diff
@@ -88,12 +124,15 @@ def prepare_virtualenv(
     execute_in_subprocess(virtualenv_cmd)
-    pip_cmd = _generate_pip_install_cmd(venv_directory, requirements)
+    pip_cmd = _generate_pip_install_cmd(
+        venv_directory, requirements, connection_id)
```
```suggestion
    pip_cmd = _generate_pip_install_cmd(venv_directory, requirements, connection_id)
```

Commit:
Syntax suggestions:
1. Docstring
remove unesessery function break of line.

ashb commented on tests/operators/test_python.py:
```diff
@@ -817,6 +837,51 @@ def f():
+                       host="localhost",
+                       schema="http",
+                       extra=json.dumps({"pypi_as_fallback": True,
+                                         "path": "/repository/python/simple"}),
```
```suggestion
                       extra={"pypi_as_fallback": True,
                              "path": "/repository/python/simple"},
```

should work now

ashb commented on tests/operators/test_python.py:
```diff
@@ -58,6 +64,17 @@
 ]
 
 
+@pytest.fixture()
+def pypi_server():
+    if not os.path.exists('/root/packages'):
+        os.makedirs('/root/packages')
+    process = subprocess.Popen(['pypi-server', '--port=8282'])
+    yield
+    process.kill()
+    shutil.rmtree('/root/packages')
+
+
+
```
Where is this `pypi-serve` command coming from?

Simpler might be to just use `python -m http.server --directory ` as that is part of the standard library

```suggestion
@pytest.fixture()
def pypi_server(tmpdir):
    process = subprocess.Popen(['python', '-m', 'http.server', '--directory', str(tmpdir),  '--port=8282'])
    yield
    process.kill()


```

Bennyelg replied:
I think this one should not changed, this due to the fact that we want to mimc the pypi server with all of its capabilities.
the pypi server is a package of course, installed for debugging (testing) only)

Commit:
remove dump

uranusjr commented on airflow/utils/python_virtualenv.py:
```diff
@@ -35,12 +38,44 @@ def _generate_virtualenv_cmd(tmp_dir: str, python_bin: str, system_site_packages
+                       'install',
+                       f'--index-url', index_url]
+
+        if pypi_as_fallback:
+            private_cmd.extend([f'--extra-index-url', 'https://pypi.org/simple'])
```
Se my comment here: https://github.com/apache/airflow/pull/16954#discussion_r669168968

If you must allow PyPI, set PyPI as the main index, not an extra index. This does not do what most people think it does.

Also I would avoid using the term “fallback” since it gives an impression PyPI is only accessed if a package is not available on the private index. That is wrong. Something like `allow_pypi` would be better (although I still strongly prefer not having this option at all).

Bennyelg replied:
In my organisation we need for example to use both, I think sometime its parameter is required. I'll change this to the allow_pypi - it seems more straight forward.

Bennyelg replied:
@uranusjr  Any way else you suggest to do it ?

uranusjr replied:
In general, when your organisaion needs `--extra-index-url`, you don’t really need it and are identifying the issue wrong 🙂 There are exceptions, e.g. you are deploying to exotic hardware that PyPI doesn’t have wheels for, and you want to supply wheels for those PyPI project. For those exceptional use cases, setting `--extra-index-url` (while keeping the main index URL pointing to PyPI) is the right thing to do. But if what you need is to install a package not on PyPI, `--extra-index-url` is the wrong solution.

The suggestion is, instead of having multiple indexes serving different packages, making all requests go through the private index server you control. The index is configured in a way that when your private package is requeted, serve your own wheels, and when a PyPI package is requested, forward the request to the actual PyPI. This mechanism is implemented by Python index tools such as devpi (called index inheritance). I heard Artifactory can do that as well, although I’ve never personally understood how Artifactory works and can’t confirm.

Bennyelg replied:
@uranusjr  I m pretty sure its safe 'as implemented' now since its not explict set as 'on' so this un"healty" behaviour will take place.

uranusjr replied:
I don’t quite understand your last message, but what I can say is the PR, as it currently stands, does not provide a good approach to set `--extra-index-url`, and I would like Airflow to not encourage the approach.
